server/src/main/java/org/opensearch/identity/scopes/Scope.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.scopes;
+
+import java.security.Principal;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.opensearch.action.ActionScope;
+import org.opensearch.extensions.ExtensionsManager;
+import org.opensearch.identity.ApplicationScope;
+import org.opensearch.plugins.ExtensionPointScope;
+
+/**
+ * Limitation for the scope of an application in OpenSearch
+ *
+ * @opensearch.experimental
+ */
+public interface Scope {
+    ScopeEnums.ScopeNamespace getNamespace();
+
+    ScopeEnums.ScopeArea getArea();
+
+    String getAction();
+
+    default String asPermissionString() {
+        return getNamespace() + "." + getArea().toString() + "." + getAction();
+    }
+
+    static Scope parseScopeFromString(String scopeAsString) {
+
+        String[] parts = scopeAsString.split("\\.");
+        if (parts.length != 3) {
+            throw new IllegalArgumentException("Invalid scope format: " + scopeAsString);
+        }
+        ScopeEnums.ScopeNamespace scopeNamespace = ScopeEnums.ScopeNamespace.fromString(parts[0]);
+        ScopeEnums.ScopeArea scopeArea = ScopeEnums.ScopeArea.fromString(parts[1]);
+        String action = parts[2];
+
+        switch (scopeNamespace) {
+            case ACTION:
+                switch (action) {
+                    case "ALL":
+                        return ActionScope.ALL;
+                    case "READ":
+                        return ActionScope.READ;
+                    default:
+                        throw new UnknownScopeException(scopeAsString);
+                }
+            case APPLICATION:
+                if (action.equals("ALL")) {
+                    return ApplicationScope.SuperUserAccess;
+                }
+                throw new UnknownScopeException(scopeAsString);
+            case EXTENSION_POINT:
+                if (action.equals("ACTION")) {
+                    return ExtensionPointScope.ACTION;
+                }
+                throw new UnknownScopeException(scopeAsString);
+            default:
+                throw new UnknownScopeException(scopeAsString);
+        }
+    }
+
+    default boolean isScopeInNamespace(String scope) {
+
+        return parseScopeFromString(scope).getNamespace().equals(getNamespace());
+    }
+
+    static Set<String> getApplicationScopes(Principal principal) {
```

Comment:
`lookup(Application) and return Scopes` seems like it shouldn't be the responsibility of the Scope, IdentityService, or ExtensionManager, it seems like there should be a (new) ApplicationManager interface + class that is aware of different applications and can be queried to get the scopes of these applications.

We can start the implementation to be very primitive and evolve it over time to better reason about plugins / extensions / etc

Reply:
I am not sure I understand the difference between an ApplicationManager and an IdentityService...

In my mind, if we have an IdentityService, it would be the thing tracking the unique identities present in the cluster whether they be applications or user. It would be able to resolve the status of the subjects and determine their scopes based on a look up from a store (probably a system index). 

I am not sure what role an ApplicationManager interface would have that was not overalling with the IndetityService interface -- what do you think?

Reply:
Update: I take it back. 

We should add another interface so that we can have an extension Manager, Application Manager, user Manager etc. 

I need to register users as applications but calling them extensions seems hacky.